test/libsolutil/CommonIO.cpp

```diff
@@ -98,6 +118,426 @@ BOOST_AUTO_TEST_CASE(readBytes_past_end)
+	BOOST_TEST(absolutePath("./../a/../b/..", reference) == absolutePathJoin(parent, ""));
+
+	BOOST_TEST(absolutePath("./a/../..", reference) == absolutePathJoin(parent, ""));
+	BOOST_TEST(absolutePath("./a/b/../../..", reference) == absolutePathJoin(parent, ""));
+	BOOST_TEST(absolutePath("./a/../b/../..", reference) == absolutePathJoin(parent, ""));
+	BOOST_TEST(absolutePath("./a/../../b/..", reference) == absolutePathJoin(parent, ""));
+
+	BOOST_TEST(absolutePath("./a/../../c", reference) == absolutePathJoin(parent, "c"));
+	BOOST_TEST(absolutePath("./a/b/../../../c", reference) == absolutePathJoin(parent, "c"));
+	BOOST_TEST(absolutePath("./a/../b/../../c", reference) == absolutePathJoin(parent, "c"));
+	BOOST_TEST(absolutePath("./a/../../b/../c/d", reference) == absolutePathJoin(parent, "c/d"));
+
+	BOOST_TEST(absolutePath("..//a", reference) == absolutePathJoin(parent, "a"));
+	BOOST_TEST(absolutePath("..//a//", reference) == absolutePathJoin(parent, "a"));
+	BOOST_TEST(absolutePath(".//a//..//..//b", reference) == absolutePathJoin(parent, "b"));
+}
+
+BOOST_AUTO_TEST_CASE(absolutePath_relative_import_with_backtracking_multiple_levels_level_into_reference)
+{
+	BOOST_TEST(absolutePath("..", "x/y/z/") == "x/y");
+	BOOST_TEST(absolutePath("../..", "x/y/z/") == "x");
+	BOOST_TEST(absolutePath("../../..", "x/y/z/") == "");
+	BOOST_TEST(absolutePath("../../../..", "x/y/z/") == "");
+
+	BOOST_TEST(absolutePath("..", "/x/y/z/") == "/x/y");
+	BOOST_TEST(absolutePath("../..", "/x/y/z/") == "/x");
+	BOOST_TEST(absolutePath("../../..", "/x/y/z/") == "/");
+	BOOST_TEST(absolutePath("../../../..", "/x/y/z/") == "");
+
+	BOOST_TEST(absolutePath("../../../../a", "x/y/z/") == "a");
+	BOOST_TEST(absolutePath("../../../../a/..", "x/y/z/") == "");
+	BOOST_TEST(absolutePath("../../../../a/../a", "x/y/z/") == "a");
+
+	BOOST_TEST(absolutePath("../.", "x/y/z/") == "x/y");
+	BOOST_TEST(absolutePath(".././..", "x/y/z/") == "x");
+	BOOST_TEST(absolutePath(".././.././..", "x/y/z/") == "");
+	BOOST_TEST(absolutePath("./../././.././.././..", "x/y/z/") == "");
+}
+
+BOOST_AUTO_TEST_CASE(absolutePath_relative_import_and_relative_segments_in_reference)
+{
+	BOOST_TEST(absolutePath(".", "x/../y/./z/") == "x/../y/./z");
+	BOOST_TEST(absolutePath("..", "x/../y/./z/") == "x/../y/.");
+	BOOST_TEST(absolutePath("../..", "x/../y/./z/") == "x/../y");
+	BOOST_TEST(absolutePath("../../..", "x/../y/./z/") == "x/..");
+	BOOST_TEST(absolutePath("../../../..", "x/../y/./z/") == "x");
+	BOOST_TEST(absolutePath("../../../../..", "x/../y/./z/") == "");
+	BOOST_TEST(absolutePath("../../../../../..", "x/../y/./z/") == "");
+}
+
+BOOST_AUTO_TEST_CASE(absolutePath_relative_import_and_consecutive_slashes_in_reference)
+{
+	BOOST_TEST(absolutePath(".", "/x/") == "/x");
+	BOOST_TEST(absolutePath("..", "/x/") == "/");
+	BOOST_TEST(absolutePath("../..", "/x/") == "");
+
+	BOOST_TEST(absolutePath(".", "//x//") == "//x//");
+	BOOST_TEST(absolutePath("..", "//x//") == "//x");
```

Comment:
Great. This is already broken on Linux, before CI even got to Windows tests (where I expect total carnage).

On my system (Arch Linux, Boost 1.81.0-7) this produces `//x`. In CI it's `//x/`.

While it's not a big deal when using an import callback because it goes to the actual filesystem, in Standard JSON it will be a problem. `//x` and `//x/` are two distinct keys.